jackson-apt-processor/src/main/java/org/dominokit/jacksonapt/processor/AbstractJsonMapperGenerator.java

```diff
@@ -166,10 +177,94 @@ private MethodSpec targetTypeMethod() {
                     enclosedFields.addAll(0, orderedProperties);
                 });
 
-        enclosedFields.addAll(getOrderedFields((TypeElement) typeUtils.asElement(superclass)));
-        return enclosedFields;
+        // Map fields type parameters to type arguments by using wrapper element
+        List<Element> orderedFields = enclosedFields
+        	.stream()
+        	.map(field -> 
+        		typeParameters.contains(typeUtils.asElement(field.asType()))? 
+        			wrapElementWithType(field, typeArguments.get(typeParameters.indexOf(typeUtils.asElement(field.asType())))):
+        			field)
+        	.collect(Collectors.toList());
+        
+        List<? extends TypeMirror> superTypeArguments  = Collections.emptyList();
+        List<? extends TypeParameterElement> superTypeParameters = Collections.emptyList();
+        if (superclass instanceof DeclaredType) {
+        	superTypeArguments = ((DeclaredType)superclass).getTypeArguments();
+        	superTypeParameters = ((TypeElement) typeUtils.asElement(superclass)).getTypeParameters();
+        }
+        
+        // Map type arguments from base type to super type arguments
+        superTypeArguments = superTypeArguments.stream()
+        	.map(typeArgument -> 
+    			typeParameters.contains(typeUtils.asElement(typeArgument))?
+    				typeArguments.get(typeParameters.indexOf(typeUtils.asElement(typeArgument)))
+    				:typeArgument)
+        	.collect(Collectors.toList());
+    					
+        orderedFields.addAll(
+        	getOrderedFields(
+        		(TypeElement) typeUtils.asElement(superclass),
+        		superTypeParameters,
+        		superTypeArguments));
+        
+        return orderedFields;
     }
-
+    
+    private Element wrapElementWithType(Element element, TypeMirror typeMirror) {
```

Comment:
@vegegoku That is very good question indeed! I would like to explain why we need such gymnastic here. 

Let's consider oderderedFields() and getOrderedFields() methods first. They are intended to scan through class hierarchy and enumerate all fields. If we want to deal with generic classes, things get more complicated, because we need to track type parameters and corresponding actual type arguments through the whole class hierarchy. One inherent problem here is that if we use typeUtils.asElement() to convert from TypeMirror to TypeElement, we loose information about actual type arguments i.e. 
`  TypeElement typeElement = (TypeElement) typeUtils.asElement(beanType);`

Note that beanType knows the actual type arguments, while typeElement doesn't! It just knows about it's type parameters.

This is the reason I needed to extend signature of getOrderedFields(...) to pass both: type parameters and type arguments.

Now let's consider enumeration of the fields inside getOrderedFields(...). Code extracts all enclosed fields by using:
`typeElement.getEnclosedElements()`
We have all enclosed fields with its corresponding types. But if the field's type is a type parameter, we do not know the actual type argument...

A short example:
`class MyTest<P> {
    P myField;
}`
`@JsonMapper
interface MyTestMapper MAPPER extends ObjectMapper<MyTest<Integer>> {}`

once we enumerate enclosed fields for MyTest<Integer> by TypeElement.getEnclosedFields(), we end up with myField with type P. We need to "replace" P with actual type argument, which in this example is Integer. Why? Because SerializerBuilder and DeserializerBuilder needs the element and its actual type to generate appropriate serializer/deserializer. Therefore, my solution is to create an element, which pretends to have type equals to actual type argument.

I guess the wrapper is not the most elegant solution, but is does a fine job. I am open for any better ideas here...

Reply:
Cant we just cast to `DeclaredType` then map each type argument from the mapper interface with the type argument in the base class?